### DISCUSSION TITLE
SDXL Training, prediction_type = 'epsilon' + snr_gamma + rescale_betas_zero_snr will case NaN

### DISCUSSION THREAD

**Discussion by Jannchie** [bghira/SimpleTuner#1241] [Q&A] [answered]
I read the document at 

https://huggingface.co/docs/diffusers/api/schedulers/ddim. 

It mentions that using `rescale_betas_zero_snr` with `v-prediction` might be more scientifically sound. However, `rescale_betas_zero_snr` with `e-prediction` should also work. Yet, when `rescale_betas_zero_snr` is `True`, the calculated snr is 0 at timestamp 999.  

And when `snr_gamma` is enabled, calculating the loss requires dividing by snr, which triggers a division by zero error, causing the loss to become `NaN`.  (https://github.com/huggingface/diffusers/blob/6dfaec348780c6153a4cfd03a01972a291d67f82/examples/text_to_image/train_text_to_image_sdxl.py#L1147)

I think `e-prediction`, `snr_gamma`, and r`escale_betas_zero_snr` should be compatible options that can work together. I don’t have much knowledge about how these work, but maybe adding a small epsilon when calculating snr could solve the problem, though I’m not sure if it’s reasonable. Could someone offer some advice?

**Comment by bghira** [comment, marked as answer]
it must be v_prediction for zsnr

**Reply by Jannchie** [reply]
Thank you for the quick response on Christmas Eve. 

Actually, I am learning based on the parameter documentation of this repository. 

In the documentation related to `rescale_betas_zero_snr`, it is stated that using `v-prediction` is “recommended.” 

https://github.com/bghira/SimpleTuner/blob/c6cdbb07cefb737a5661db99b974d16d96db8890/OPTIONS.md?plain=1#L853-L857

So I mistakenly assumed that it would work correctly under e-prediction as well. 

Perhaps it might be necessary to raise an error when `rescale_betas_zero_snr` is set to `True` and the prediction type is not `v-prediction`?

**Reply by bghira** [reply]
nope, because some people insist that it works for them, and they complain when these limits are put into place.
